[Gradient Compression] Refactor CommHookInterface and PythonCommHook.

wayi1 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #46566 [Gradient Compression] Add CppCommHook subclass for supporting the C++ API of communication hook.
* **#46512 [Gradient Compression] Refactor CommHookInterface and PythonCommHook.**

1. Merge 1-line PythonCommHook constructor into the header for simplicity.
2. Move the implementation of PythonCommHook destructor from the header file to cpp file.
3. Rename processFuture method as parseHookResult for readability.
4. Simplify some comments.

Original PR issue: C++ DDP Communication Hook https://github.com/pytorch/pytorch/issues/46348

Differential Revision: [D24374282](https://our.internmc.facebook.com/intern/diff/D24374282/)

Commit:
[Gradient Compression] Refactor CommHookInterface and PythonCommHook.

1. Let CommHookInterface::processFuture be a non-virtual method, which can be shared by both Python and C++ implementations.
2. Merge 1-line PythonCommHook constructor into the header for simplicity.
3. Rename processFuture method as parseFromHookResult for readability.
4. Simplify the comments.

Original PR issue: C++ DDP Communication Hook https://github.com/pytorch/pytorch/issues/46348

Differential Revision: [D24374282](https://our.internmc.facebook.com/intern/diff/D24374282/)

[ghstack-poisoned]

Commit:
Update on "[Gradient Compression] Refactor CommHookInterface and PythonCommHook."


1. Let CommHookInterface::processFuture be a non-virtual method, which can be shared by both Python and C++ implementations.
2. Merge 1-line PythonCommHook constructor into the header for simplicity.
3. Rename processFuture method as parseFromHookResult for readability.
4. Simplify the comments.

Original PR issue: C++ DDP Communication Hook https://github.com/pytorch/pytorch/issues/46348

Differential Revision: [D24374282](https://our.internmc.facebook.com/intern/diff/D24374282/)

[ghstack-poisoned]

pritamdamania87 commented on torch/csrc/distributed/c10d/comm.h:
```diff
@@ -35,45 +35,33 @@ class GradBucket {
-  // parameters.
-  virtual std::vector<at::Tensor> processFuture(c10::IValue future_value) = 0;
+  // Returns the resulting tensors once the communication hook result is ready.
+  std::vector<at::Tensor> parseFromHookResult(const c10::IValue& result);
```
nit: `parseResult` or `parseHookResult` might be sufficient here.

SciPioneer replied:
Used "parseHookResult". Usually I prefer the naming style of either parseToXXX or parseFromXXX, but I guess parseXXX here should be clear.

pritamdamania87 commented on torch/csrc/distributed/c10d/comm.h:
```diff
@@ -35,45 +35,33 @@ class GradBucket {
-  // bucket's tensors once ready.
+  // Runs the registered communication hook to communicate gradients
+  // asynchronously, Returns a future that holds the communication results.
   virtual c10::intrusive_ptr<torch::jit::Future> runHook(
```
Should we preserve the old comments as well? They provide useful information about the hook's contract?

SciPioneer replied:
Feel the old comments are a bit wordy, so rephrased it from 5 lines to 3 lines, which should not really lose any information.

pritamdamania87 commented on torch/csrc/distributed/c10d/comm.h:
```diff
@@ -35,45 +35,33 @@ class GradBucket {
 
-  // Once the grad bucket of Future is ready, c10d reducer will call this
-  // function to get the resulting tensors of the grad bucket. Then c10d
-  // reducer will use these tensors and copy grads to the grads of individual
```
We should preserve these old comments as well indicating how the hook is used by reducer.

pritamdamania87 commented on torch/csrc/distributed/c10d/comm.h:
```diff
@@ -35,45 +35,33 @@ class GradBucket {
-  // maintain and update any state information that users would like to maintain
-  // as part of the training process. The hook can perform whatever processing
-  // user specified and return a Future indicating completion of any async work.
-  PythonCommHook(py::object state, py::object hook);
```
Preserve old comments.

pritamdamania87 commented on torch/csrc/distributed/c10d/comm.cpp:
```diff
@@ -85,8 +85,23 @@ void broadcast_coalesced(
 
-PythonCommHook::PythonCommHook(py::object state, py::object hook)
-    : state_(std::move(state)), hook_(std::move(hook)){};
+std::vector<at::Tensor> CommHookInterface::parseFromHookResult(
+    const c10::IValue& result) {
+  TORCH_INTERNAL_ASSERT(
+      result.isPyObject() || result.isTensorList(),
+      "expected the future value is either a PyObject (if PythonCommHook is used) or TensorList (if CppCommHook is used)");
+
+  if (result.isPyObject()) {
+    py::gil_scoped_acquire ag;
+    py::object obj = torch::jit::toPyObject(result);
+    auto value = torch::jit::toIValue(
+        obj, c10::ListType::create(c10::TensorType::get()));
+
+    return value.toTensorVector();
+  }
+
+  return result.toTensorVector();
+}
```
This should be a pure virtual method in CommHookInterface such that both PythonCommHook and CppCommHook have their own implementations.

A subclass of CommHookInterface can override runHook to return an IValue of any type (ex: List/Tuple etc.). It's probably best for the subclass to handle the return value and not in the base class itself.

Commit:
Update on "[Gradient Compression] Refactor CommHookInterface and PythonCommHook."


1. Let CommHookInterface::processFuture be a non-virtual method, which can be shared by both Python and C++ implementations.
2. Merge 1-line PythonCommHook constructor into the header for simplicity.
3. Rename processFuture method as parseFromHookResult for readability.
4. Simplify the comments.

Original PR issue: C++ DDP Communication Hook https://github.com/pytorch/pytorch/issues/46348

Differential Revision: [D24374282](https://our.internmc.facebook.com/intern/diff/D24374282/)

[ghstack-poisoned]

Commit:
Update on "[Gradient Compression] Refactor CommHookInterface and PythonCommHook."


1. Merge 1-line PythonCommHook constructor into the header for simplicity.
2. Move the implementation of PythonCommHook destructor from the header file to the cpp file.
3. Rename processFuture method as parseHookResult for readability.
4. Simplify some comments.

Original PR issue: C++ DDP Communication Hook https://github.com/pytorch/pytorch/issues/46348

Differential Revision: [D24374282](https://our.internmc.facebook.com/intern/diff/D24374282/)

[ghstack-poisoned]

pritamdamania87 commented on torch/csrc/distributed/c10d/comm.cpp:
```diff
@@ -85,8 +85,16 @@ void broadcast_coalesced(
 
-PythonCommHook::PythonCommHook(py::object state, py::object hook)
-    : state_(std::move(state)), hook_(std::move(hook)){};
+PythonCommHook::~PythonCommHook() {
+  py::gil_scoped_acquire ag;
+  state_.dec_ref();
+  hook_.dec_ref();
+  // Explicitly set state_ and hook_ to nullptr to prevent py::object's dtor
+  // to decref on the PyObject again.
+  // See Note [Destructing py::object] in python_ivalue.h
+  state_.ptr() = nullptr;
+  hook_.ptr() = nullptr;
+}
```
Thanks for fixing this, did you see some failures that were caused since this was missing?

SciPioneer replied:
Thanks for catching this!

Realized that there is no easy way to fix this, so I reverted this change. Realized that the definition of both constructor and destructor of a TORCH_API class should only be used in libtorch_core target in build_variables.bzl, but this cpp file belongs to libtorch_python_distributed_core_sources, which is an extension of libtorch_python_sources.

See this picture:
![image](https://user-images.githubusercontent.com/8884619/96962443-a1bead80-14bb-11eb-8a1d-011475616d2d.png)

On the other hand, not sure if there can be a workaround by somehow extending the file WindowsTorchApiMacro.h (e.g., adding a new specifier like TORCH_CPP_API) @mrshenli  is this possible?

SciPioneer replied:
It seems that reverting this change still couldn't fix the compiler error on Windows platform.

SciPioneer replied:
Replaced TORCH_API by TORCH_PYTHON_API for CommHookInterface.

pritamdamania87 reviewed (commented):
Code changes look good, although it seems like the windows tests are failing with an error like:

```
comm.cpp.obj : error LNK2019: unresolved external symbol "__declspec(dllimport) public: virtual __cdecl c10d::CommHookInterface::~CommHookInterface(void)" (__imp_??1CommHookInterface@c10d@@UEAA@XZ) referenced in function "public: virtual __cdecl c10d::PythonCommHook::~PythonCommHook(void)" (??1PythonCommHook@c10d@@UEAA@XZ)
init.cpp.obj : error LNK2019: unresolved external symbol "__declspec(dllimport) public: __cdecl c10d::CommHookInterface::CommHookInterface(void)" (__imp_??0CommHookInterface@c10d@@QEAA@XZ) referenced in function "class std::unique_ptr<class c10d::PythonCommHook,struct std::default_delete<class c10d::PythonCommHook> > __cdecl std::make_unique<class c10d::PythonCommHook,class pybind11::object,class pybind11::object,0>(class pybind11::object &&,class pybind11::object &&)" (??$make_unique@VPythonCommHook@c10d@@Vobject@pybind11@@V34@$0A@@std@@YA?AV?$unique_ptr@VPythonCommHook@c10d@@U?$default_delete@VPythonCommHook@c10d@@@std@@@0@$$QEAVobject@pybind11@@0@Z)
bin\torch_python.dll : fatal error LNK1120: 2 unresolved externals
```

Commit:
Update on "[Gradient Compression] Refactor CommHookInterface and PythonCommHook."


1. Merge 1-line PythonCommHook constructor into the header for simplicity.
2. Move the implementation of PythonCommHook destructor from the header file to the cpp file.
3. Rename processFuture method as parseHookResult for readability.
4. Simplify some comments.

Original PR issue: C++ DDP Communication Hook https://github.com/pytorch/pytorch/issues/46348

Differential Revision: [D24374282](https://our.internmc.facebook.com/intern/diff/D24374282/)

[ghstack-poisoned]

Commit:
Update on "[Gradient Compression] Refactor CommHookInterface and PythonCommHook."


1. Merge 1-line PythonCommHook constructor into the header for simplicity.
2. Move the implementation of PythonCommHook destructor from the header file to the cpp file.
3. Rename processFuture method as parseHookResult for readability.
4. Simplify some comments.

Original PR issue: C++ DDP Communication Hook https://github.com/pytorch/pytorch/issues/46348

Differential Revision: [D24374282](https://our.internmc.facebook.com/intern/diff/D24374282/)

[ghstack-poisoned]

Commit:
Update on "[Gradient Compression] Refactor CommHookInterface and PythonCommHook."


1. Merge 1-line PythonCommHook constructor into the header for simplicity.
2. Rename processFuture method as parseHookResult for readability.
3. Simplify some comments.

Original PR issue: C++ DDP Communication Hook https://github.com/pytorch/pytorch/issues/46348

Differential Revision: [D24374282](https://our.internmc.facebook.com/intern/diff/D24374282/)

[ghstack-poisoned]

Commit:
Update on "[Gradient Compression] Refactor CommHookInterface and PythonCommHook."


1. Merge 1-line PythonCommHook constructor into the header for simplicity.
2. Rename processFuture method as parseHookResult for readability.
3. Simplify some comments.

Original PR issue: C++ DDP Communication Hook https://github.com/pytorch/pytorch/issues/46348

Differential Revision: [D24374282](https://our.internmc.facebook.com/intern/diff/D24374282/)

[ghstack-poisoned]

Commit:
Update on "[Gradient Compression] Refactor CommHookInterface and PythonCommHook."


1. Merge 1-line PythonCommHook constructor into the header for simplicity.
2. Rename processFuture method as parseHookResult for readability.
3. Simplify some comments.

Original PR issue: C++ DDP Communication Hook https://github.com/pytorch/pytorch/issues/46348

Differential Revision: [D24374282](https://our.internmc.facebook.com/intern/diff/D24374282/)

[ghstack-poisoned]

Commit:
Update on "[Gradient Compression] Refactor CommHookInterface and PythonCommHook."


1. Merge 1-line PythonCommHook constructor into the header for simplicity.
2. Rename processFuture method as parseHookResult for readability.
3. Simplify some comments.

Original PR issue: C++ DDP Communication Hook https://github.com/pytorch/pytorch/issues/46348

Differential Revision: [D24374282](https://our.internmc.facebook.com/intern/diff/D24374282/)

[ghstack-poisoned]

Commit:
Update on "[Gradient Compression] Refactor CommHookInterface and PythonCommHook."


1. Merge 1-line PythonCommHook constructor into the header for simplicity.
2. Rename processFuture method as parseHookResult for readability.
3. Simplify some comments.

Original PR issue: C++ DDP Communication Hook https://github.com/pytorch/pytorch/issues/46348

Differential Revision: [D24374282](https://our.internmc.facebook.com/intern/diff/D24374282/)

[ghstack-poisoned]

Commit:
Update on "[Gradient Compression] Refactor CommHookInterface and PythonCommHook."


1. Merge 1-line PythonCommHook constructor into the header for simplicity.
2. Move the implementation of PythonCommHook destructor from the header file to cpp file.
3. Rename processFuture method as parseHookResult for readability.
4. Simplify some comments.

Original PR issue: C++ DDP Communication Hook https://github.com/pytorch/pytorch/issues/46348

Differential Revision: [D24374282](https://our.internmc.facebook.com/intern/diff/D24374282/)

[ghstack-poisoned]
